### PR TITLE
[WIP] Add AES/HMAC demo application

### PR DESCRIPTION
**Pull request by vogelpi**
I hacked together a simple example app that reads some plain text and key from the UART, then encodes/decodes this text, and prints it over UART.

This is work in progress and not ready for review. I made the PR mainly to make it available for someone willing to support me on the HMAC side.

This is based on my PR #1129.

### CODE REVIEW

**Comment by sjgitty** [comment]
sounds like @moidx has offered to contribute top-level demo code, and
@eunchan is willing to help with the HMAC side. Thanks guys and thanks
for starting this off @vogelpi

File: sw/device/demos/aes_hmac/aes_hmac_demo.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+void aes_process_blocks(void *data_out, const void *data_in, int num_blocks) {
+  uint32_t *word_ptr_out = (uint32_t *)data_out;
+  const uint32_t *word_ptr_in = (uint32_t *)data_in;
+
+  if (num_blocks > 0) {
+    // Write Input Data Block 0
+    aes_data_put_wait((const void *)word_ptr_in);
+
+    if (num_blocks > 1) {
+      // Write Input Data Block 0
```
Comment (sjgitty) [inline sw/device/demos/aes_hmac/aes_hmac_demo.c]: Block 1?

File: sw/device/demos/aes_hmac/aes_hmac_demo.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+#include "sw/device/lib/aes.h"
+#include "sw/device/lib/common.h"
+#include "sw/device/lib/uart.h"
+
+#define LENGTH_KEY 32
+#define MAX_LENGTH_TEXT 128
+
+/**
+ * Wait for AES unit to have valid output data and then
```
Comment (sjgitty) [inline sw/device/demos/aes_hmac/aes_hmac_demo.c]: this comment seems out of place with the functionality of this routine

File: sw/device/demos/aes_hmac/aes_hmac_demo.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  bool done = false;
+  int ret = -1;
+  char my_char = 0;
+  int index = 0;
+
+  while (done == false) {
+    ret = uart_rcv_char(&my_char);
+
+    if (ret == 0) {
+      if (my_char == 13) {
```
Comment (sjgitty) [inline sw/device/demos/aes_hmac/aes_hmac_demo.c]: maybe a comment that `13` (or `0x0d`) is the carriage return

**Comment by moidx** [comment]
Quickly added some hashing to the demo.  Pushed changes to eval proposal. The demo currently does the following: 

1. digest_initial = SHA256(plaintext) and print it.
2. ciphertext = AES(key, plaintext), and print it.
3. plaintext = AES(key, ciphertext), and print it.
4. digest_final = SHA256(plaintext), and print error if digest_initial != digest_final

Is this sufficient? s/SHA256/HMAC/ is easy to switch but I figured SHA was good enough for the purposes of this demo.

Here is a follow up idea... not sure if overkill, but let me know what you think: 

If we want to make it more interactive, we could add some host to device communication, and have the host calculate a payload for OpenTitan to decrypt and check integrity before doing something.

* host --> (send encrypted blob + hmac) --> OpenTitan
* OpenTitan decrypt + hmac check. If valid action do something fun, else do something fun too
* OpenTitan --> (send encrypted blob + hmac) --> host
* Host checks response

This would require a shared secret between host and OpenTitan. wdyt?

File: sw/device/demos/aes_hmac/aes_hmac_demo.c (hunk 1 above)
Comment (vogelpi) [reply sw/device/demos/aes_hmac/aes_hmac_demo.c]: Done.

File: sw/device/demos/aes_hmac/aes_hmac_demo.c (hunk 2 above)
Comment (vogelpi) [reply sw/device/demos/aes_hmac/aes_hmac_demo.c]: Done.

File: sw/device/demos/aes_hmac/aes_hmac_demo.c (hunk 3 above)
Comment (vogelpi) [reply sw/device/demos/aes_hmac/aes_hmac_demo.c]: Done.

**Comment by vogelpi** [comment]
HI @moidx,
thanks a lot for adding the SHA/HMAC-part.

Some host to device communication would be really great. Using known software libraries like OpenSSL/BoringSSL on the host side, we could demonstrate that our accelerators are real. Right now, since the same unit does the encode/decode it is less obvious that the implementation follows the specification/standard (theoretically, it just needs to be compatible to itself for this example).

However, involving also the host will make testing much more complex. At the moment, I do not have an FPGA board at hand. Or do you think this could also be done with Verilator?

**Comment by eunchan** [comment]
Can I get which FPGA version you are using? I am getting error as below:

```console
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/spi_device/u_clk_spi/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.
WARNING: [Power 33-230] Invalid input clock frequency for top_earlgrey/u_dm_top/dap/i_dmi_jtag_tap/i_tck_inv/gen_scan.i_dft_tck_mux/gen_xilinx.u_impl_xilinx/bufgmux_i. Out of range!
Resolution: Please refer to Clocking Resources User Guide for valid input clock frequency.

Finished Running Vector-less Activity Propagation
INFO: [Pwropt 34-9] Applying IDT optimizations ...
INFO: [Pwropt 34-10] Applying ODC optimizations ...


Starting PowerOpt Patch Enables Task
INFO: [Pwropt 34-162] WRITE_MODE attribute of 0 BRAM(s) out of a total of 146 has been updated to save power. Run report_power_opt to get a complete listing of the BRAMs updated.
INFO: [Pwropt 34-201] Structural ODC has moved 0 WE to EN ports
Number of BRAM Ports augmented: 80 newly gated: 0 Total Ports: 292
Ending PowerOpt Patch Enables Task | Checksum: 155850271

Time (s): cpu = 00:00:00.43 ; elapsed = 00:00:00.44 . Memory (MB): peak = 3047.512 ; gain = 0.000 ; free physical = 13520 ; free virtual = 60458
Ending Power Optimization Task | Checksum: 155850271

Time (s): cpu = 00:00:47 ; elapsed = 00:00:23 . Memory (MB): peak = 3047.512 ; gain = 624.645 ; free physical = 13558 ; free virtual = 60495

Starting Final Cleanup Task
Ending Final Cleanup Task | Checksum: 155850271

Time (s): cpu = 00:00:00 ; elapsed = 00:00:00 . Memory (MB): peak = 3047.512 ; gain = 0.000 ; free physical = 13558 ; free virtual = 60495

Starting Netlist Obfuscation Task
Netlist sorting complete. Time (s): cpu = 00:00:00.04 ; elapsed = 00:00:00.04 . Memory (MB): peak = 3047.512 ; gain = 0.000 ; free physical = 13558 ; free virtual = 60495
Ending Netlist Obfuscation Task | Checksum: 15e173789

Time (s): cpu = 00:00:00.04 ; elapsed = 00:00:00.04 . Memory (MB): peak = 3047.512 ; gain = 0.000 ; free physical = 13560 ; free virtual = 60497
INFO: [Common 17-83] Releasing license: Implementation
33 Infos, 25 Warnings, 11 Critical Warnings and 0 Errors encountered.
opt_design completed successfully
opt_design: Time (s): cpu = 00:01:05 ; elapsed = 00:00:37 . Memory (MB): peak = 3047.512 ; gain = 1296.371 ; free physical = 13560 ; free virtual = 60497
source /build/lowrisc_systems_top_earlgrey_nexysvideo_0.1/synth-vivado/vivado_hook_opt_design_post.tcl
INFO: [Designcheck 2-1] Checking if ROM memory is mapped to BRAM memory.
WARNING: [Vivado 12-180] No cells matched 'get_cells -hierarchical -filter { NAME =~  "*rom_rom*dout_o_reg_0" && PRIMITIVE_TYPE =~ BMEM.*.* }'.
ERROR: [Designcheck 2-3] BRAM implementation not found for ROM memory.
ERROR: [runtcl-1] ERROR: [Common 17-39] 'send_msg_id' failed due to earlier errors.
ERROR: [Common 17-39] 'send_msg_id' failed due to earlier errors.

INFO: [Common 17-206] Exiting Vivado at Thu Dec  5 10:19:28 2019...
[Thu Dec  5 10:19:28 2019] impl_1 finished
wait_on_run: Time (s): cpu = 00:04:25 ; elapsed = 00:04:30 . Memory (MB): peak = 1341.891 ; gain = 0.000 ; free physical = 14856 ; free virtual = 61793
# open_run impl_1
ERROR: [Common 17-69] Command failed: Run 'impl_1' failed. Unable to open
INFO: [Common 17-206] Exiting Vivado at Thu Dec  5 10:19:28 2019...
make: *** [Makefile:6: lowrisc_systems_top_earlgrey_nexysvideo_0.1.bit] Error 1
ERROR: Failed to build lowrisc:systems:top_earlgrey_nexysvideo:0.1 : 'make' exited with an error code
```

**Comment by moidx** [comment]
@eunchan I was testing on Verilator. Let me try the FPGA.

**Comment by moidx** [comment]
> HI @moidx,
> thanks a lot for adding the SHA/HMAC-part.
> 
> Some host to device communication would be really great. Using known software libraries like OpenSSL/BoringSSL on the host side, we could demonstrate that our accelerators are real. Right now, since the same unit does the encode/decode it is less obvious that the implementation follows the specification/standard (theoretically, it just needs to be compatible to itself for this example).
> 
> However, involving also the host will make testing much more complex. At the moment, I do not have an FPGA board at hand. Or do you think this could also be done with Verilator?

@vogelpi we can probably add a hex string based protocol on top of UART just to make it easier to run in Verilator and FPGA, as part of a pytest or in interactive mode. What do you think?

**Comment by eunchan** [comment]
> @vogelpi we can probably add a hex string based protocol on top of UART just to make it easier to run in Verilator and FPGA, as part of a pytest or in interactive mode. What do you think?

Or... ZMODEM?

**Comment by GregAC** [comment]
WRT the integrity check can we please keep it just a simple AES/HMAC demo (i.e. encrypt and hash some supplied data) without the integrity mockup as discussed in the demo coordination document?

**Comment by vogelpi** [comment]
Sorry guys, I did not manage to continue the work on this demo. Does it work on your FPGA @moidx?

Based on @GregAC 's comment we maybe best leave it as it is?

**Comment by eunchan** [comment]
On FPGA it works. My bit file wasn't able to run SPIFLASH firmware download, but 
Miguel's bitstream worked on my FPGA board and I could run the AES_HMAC demo.

Thanks Pirmin and Miguel for the work.

What we could improve on top of this PR is:

- Adding XMODEM/ZMODEM to download binary file (and run AES and/or HMAC with 
  given key)
- On the host side, give a way to compare the OT result and expected result.

**Comment by eunchan** [comment]
One more desired feature in the demo (talked with Miguel) is to have the demo run by itself with some dictionary, So that the presenter doesn't have to type to run the demo.

**Comment by vogelpi** [comment]
Thanks for your feedback @eunchan . I added a non-interactive mode (now the default). The user does not have to provide any input in this case. To enable the previous interactive mode, Switch 0 on the FPGA board needs to be in the ON position.

Regarding adding communication with the host, as @GregAC pointed out we should maybe keep it simple for this demo and not go into the direction of the integrity mockup.

I would consider this demo as being done. Should we clean it up and then merge it. Or what is the way to go with our demos @GregAC ?

**Comment by GregAC** [comment]
Thanks @vogelpi 

I think it works best in interactive mode, allows people to alter the key and see the ciphertext change and the hash remain the same etc. You can always just hit enter to avoid having to type much.

I've added a loop so it'll go round and do it again. It has the nice property of remembering what text/key you set last time.

As for whether we should merge it there is maybe some value in having little demo/test programs to demonstrate how things work. However also adds a bit of maintenance workload as they could break if we end up altering the HW spec.

**Comment by vogelpi** [comment]
Okay, great. Maybe we can merge your changes into this PR, clean it up and then merge after the summit.

**Comment by eunchan** [comment]
Should we close this PR?

**Comment by vogelpi** [comment]
Hi @eunchan,
I discussed with @GregAC who did some modifications locally. Once he pushed his modifications, I will archive it in my fork and close the PR.

**Comment by vogelpi** [comment]
Thanks @GregAC for sharing your changes. I will now close this PR. The demo remains in my fork:
https://github.com/vogelpi/opentitan-1/tree/aes-hmac-demo

**Pull request closed without merging**